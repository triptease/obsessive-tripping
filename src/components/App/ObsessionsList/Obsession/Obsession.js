import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

const ScoringButton = styled.button`
  margin-left: 4px;
  outline: none;
  ${({ voted }) => voted && 'border: 4px solid gold;'};
`

const Byline = styled.aside`
  font-size: 0.8rem;
`

class Obsession extends PureComponent {
  static propTypes = {
    title: PropTypes.string.isRequired,
    onVote: PropTypes.func,
    score: PropTypes.number,
    submitterEmail: PropTypes.string,
    submitterName: PropTypes.string,
    submitterSlackURL: PropTypes.string,
    userId: PropTypes.string
  }

  onVote = value => {
    const { onVote, onDeleteVote, id, userId, vote } = this.props
    if (!userId) {
      console.log('You need to be logged-in to vote!')
      return
    }
    if (vote !== value) {
      onVote(id, { userId, value })
    } else {
      onDeleteVote(id, { userId })
    }
  }

  onLikeClick = this.onVote.bind(undefined, 1)

  onDislikeClick = this.onVote.bind(undefined, -1)

  render() {
    const {
      score,
      title,
      submitterName,
      submitterSlackURL,
      submitterEmail,
      vote
    } = this.props
    return (
      <div>
        {title} | Score: {typeof score !== 'undefined' ? score : 'N/A'}
        <ScoringButton onClick={this.onLikeClick} voted={vote === 1}>
          <span role="img" aria-label="thumbs up">
            üëç
          </span>
        </ScoringButton>
        <ScoringButton onClick={this.onDislikeClick} voted={vote === -1}>
          <span role="img" aria-label="thumbs up">
            üëé
          </span>
        </ScoringButton>
        {submitterName ? (
          <Byline>
            Submitted by:{' '}
            <a href={submitterSlackURL || `mailto:${submitterEmail}`}>
              {submitterName}
            </a>
          </Byline>
        ) : null}
      </div>
    )
  }
}

export default Obsession
