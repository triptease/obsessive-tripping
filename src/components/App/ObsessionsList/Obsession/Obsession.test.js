import { shallow } from 'enzyme'
import React from 'react'

import Obsession, { DeleteButton } from './Obsession'

describe('Obsession', () => {
  it('renders', () => {
    const component = shallow(<Obsession id="ID0001" title="TDD" />)
    expect(component).toMatchSnapshot()
  })

  it('calls its `onDelete` with its id when the delete button is clicked', () => {
    const onDelete = jest.fn()
    const component = shallow(
      <Obsession id="ID0001" title="TDD" onDelete={onDelete} />
    )
    component.find(DeleteButton).simulate('click')
    expect(onDelete).toBeCalledWith('ID0001')

    component.setProps({ id: 'different-id' })
    component.find(DeleteButton).simulate('click')
    expect(onDelete).toBeCalledWith('different-id')
  })
})
