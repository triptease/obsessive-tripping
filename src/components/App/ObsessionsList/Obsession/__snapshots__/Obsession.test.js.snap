// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Obsession renders 1`] = `
<div>
  TDD
   | Score: 
  N/A
  <Component
    border={false}
    onClick={[Function]}
  />
  <Component
    border={false}
    onClick={[Function]}
  />
  <Component
    onClick={[Function]}
  />
</div>
`;
